Below is a demonstration of using TensorFlow to classify digits from the MNIST dataset.

In [1]:
import sys
import tensorflow as tf

# Display environment versions
print(f"Python version: {sys.version}")
print(f"TensorFlow version: {tf.__version__}")

Python version: 3.12.13 (main, Mar  4 2026, 09:23:07) [GCC 11.4.0]
TensorFlow version: 2.20.0


In [2]:
mnist = tf.keras.datasets.mnist

(x_train, y_train), (x_test, y_test) = mnist.load_data()

x_train = tf.keras.utils.normalize(x_train, axis=1)
x_test = tf.keras.utils.normalize(x_test, axis=1)

model = tf.keras.models.Sequential()

model.add(tf.keras.layers.Flatten())
model.add(tf.keras.layers.Dense(64, activation=tf.keras.activations.sigmoid))
model.add(tf.keras.layers.Dense(128, activation=tf.keras.activations.sigmoid))
model.add(tf.keras.layers.Dense(10, activation=tf.keras.activations.softmax)) # 10 output classes

model.compile(optimizer="adam", loss="sparse_categorical_crossentropy", metrics=["accuracy"])

model.fit(x=x_train, y=y_train, epochs=10)

print(model.summary())

test_loss, test_acc = model.evaluate(x=x_test, y=y_test, verbose=0)
print(f'\nTest accuracy: {test_acc * 100:.2f}%')

11490434/11490434 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step
Epoch 1/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 8s 3ms/step - accuracy: 0.8367 - loss: 0.6055
Epoch 2/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 10s 3ms/step - accuracy: 0.9287 - loss: 0.2390
Epoch 3/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 5s 3ms/step - accuracy: 0.9467 - loss: 0.1800
Epoch 4/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 6s 3ms/step - accuracy: 0.9564 - loss: 0.1452
Epoch 5/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 5s 3ms/step - accuracy: 0.9636 - loss: 0.1220
Epoch 6/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 5s 3ms/step - accuracy: 0.9687 - loss: 0.1046
Epoch 7/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 6s 3ms/step - accuracy: 0.9725 - loss: 0.0901
Epoch 8/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 5s 3ms/step - accuracy: 0.9758 - loss: 0.0791
Epoch 9/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 6s 3ms/step - accuracy: 0.9788 - loss: 0.0693
Epoch 10/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 5s 3ms/step - accuracy: 0.9817 - loss: 0.0612


Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ flatten (Flatten)               │ (32, 784)              │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense (Dense)                   │ (32, 64)               │        50,240 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (32, 128)              │         8,320 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_2 (Dense)                 │ (32, 10)               │         1,290 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 179,552 (701.38 KB)

 Trainable params: 59,850 (233.79 KB)

 Non-trainable params: 0 (0.00 B)

 Optimizer params: 119,702 (467.59 KB)

None

Test accuracy: 96.90%


1. Data Loading and Preprocessing

The process begins by loading the MNIST dataset, which contains 70,000 grayscale images of numerical digits (0–9).  Loading: The data is split into a training set (used to teach the model) and a test set (used to verify its accuracy).  Normalization: The pixel values are scaled to a range between 0 and 1 using tf.keras.utils.normalize. This helps the model learn faster and more effectively by ensuring "training stability".

  2. Building the Model Architecture
  
  The model is built using a Sequential structure, where data flows through layers in order:  Flatten Layer: Converts the 2D image ($28 \times 28$ pixels) into a 1D vector of 784 values.  Hidden Layers: Two "Dense" (fully connected) layers with 64 and 128 neurons respectively. They use the Sigmoid activation function to introduce non-linearity.  Output Layer: Consists of 10 neurons, one for each digit (0–9). The Softmax activation is used here to turn the output into a probability distribution, where the highest value represents the model's prediction.  

  3. Compilation and Training
  
  Before training, the model is compiled with specific settings:  Optimizer (Adam): The algorithm that updates the weights of the neurons to reduce errors.  Loss Function (Sparse Categorical Crossentropy): Measures how "wrong" the model's predictions are.  Training: The model.fit function runs the training for 10 epochs (10 full passes through the dataset).

   4. Evaluation
   
   Finally, the model is tested on the test set—data it has never seen before. This provides the Test Accuracy, which tells us how well the model generalizes to new, real-world examples of handwriting.  